File: String/1807.Evaluate the Bracket Pairs of a String.ts
function evaluate(s: string, knowledge: string[][]): string {
    let map = new Map<string, string>();

    for (let [key, val] of knowledge) {
        map.set(key, val);
    }

    let n = s.length;
    let open = false;
    let newStr = "";
    let temp = "";


    for (let i = 0; i < n; i++) {
        if (s[i] == "(") {
            open = true;
            continue;
        }

        if (open) {
            if (s[i] == ")") {

                newStr += map.has(temp) ? map.get(temp) : "?";
                temp = "";
                open = false;
            } else {
                temp += s[i];
            }
        } else {
            newStr += s[i];
        }
    }
    return newStr;
};

File: String/1807.Evaluate the Bracket Pairs of a String.cpp
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

class Solution
{
public:
    string evaluate(string s, vector<vector<string>> &knowledge)
    {
        unordered_map<string, string> kMap;

        for (auto &pair : knowledge)
        {
            string key = pair[0];
            string value = pair[1];
            kMap[key] = value;
        }

        int n = s.length();
        string newStr;
        bool open = false;
        string temp = "";
        for (char c : s)
        {
            if (c == '(')
            {
                open = true;
                continue;
            }

            if (open)
            {
                if (c == ')')
                {
                    newStr += (kMap.find(temp) != kMap.end() ? kMap[temp] : "?");
                    temp = "";
                    open = false;
                }
                else
                {
                    temp += c;
                }
            }
            else
            {
                newStr += c;
            }
        }
        return newStr;
    }
};